blog/zk-snark-pallet.md

```diff
@@ -0,0 +1,263 @@
+
+## Starting point
+
+We have already learned that zk-SNARKs can be used to prove knowledge of a solution for a problem, without revealing it. We just need to provide proof that can be later verified by someone else. Creation of such proof is done in a couple of stages, where we first convert our problem to R1CS (*Rank-1 Constraint System*) form, and then transform it to the QAP (*Quadratic Arithmetic Program*). For this process, we used Circom and SnarkJS. We used a Groth16 as a proving system for assurance of the encryption. Finally, we were able to implement a Rust library, which used artifacts from Circom and SnarkJS, to validate proof.
+
+In the previous post, we show how we can use Circom and SnarkJS in the verification process. Let's remind us what artifacts in the context of those tools do we get:
+* **[public.json](https://github.com/bright/zk-snarks-with-substrate/blob/M2-post/blog/data/public.json)** - this file contains public input, in our case, it is a `12` value from our equation.
+* **[verification_key.json](https://github.com/bright/zk-snarks-with-substrate/blob/M2-post/blog/data/verification_key.json)** - file generated from SnarkJS, it contains a verification key, which "signs" our circuits (transformed equation).
+* **[proof.json](https://github.com/bright/zk-snarks-with-substrate/blob/M2-post/blog/data/proof.json)** - file created by Alice using SnarkJS, which proves her knowledge of solving the equation. 
+
+As you probably remember, at this stage Bob could use a SnarkJS to verify, if the proof is valid, by running this command:
+```
+snarkjs groth16 verify verification_key.json public.json proof.json
+```
+He is using a *Groth16* as a proving system. The output is:
+```
+[INFO]  snarkJS: OK!
+```
+which means that that proof passed the validation. Now we will try to do the same in the Substrate pallet, using our Groth16 code from the previous article.
+
+
+## Substrate Pallet
+First, let's define what we expect from pallet to do. We definitely would like to have on-chain storage of the artifacts generated from Circom and SnarkJS, because they are going to be used in the verification process. We will need to provide functionality for other participants to send and verify their proofs. Finally, we would like to be informed when someone will send us valid proof.
+
+Based on what we said, we can define an interface for our pallet which is in Substrate called an extirices. We are going to define two methods:
+* **setup_verification** - this methods allows Bob to send a public inputs (*public.json*) and the verification key (*verification_key.json*).
+* **verify** - thanks to this method, Alice (and others) will be able to send their proofs (*proof.json*) and verify them. When the verification succeeded, it will emit an event that is going to be stored on the blockchain.
```

Comment:
Tu bym dałaI conditional, czyli: When the verification succeeds, it will emit an event that is going to be stored on the blockchain.